internal/cache/redis.go

```diff
@@ -0,0 +1,174 @@
+			DB:       config.Application.Cache.Redis.DB,
+		})
+	} else {
+		log.Logger.Info().Bool("cache", true).Msg("No application cache in use")
+	}
+}
+
+func FindAppTypeId(_ context.Context) (string, error) {
+	appTypeIdMutex.Lock()
+	defer appTypeIdMutex.Unlock()
+
+	if appTypeId == nil {
+		return "", ErrNotFound
+	}
+	return *appTypeId, nil
+}
+
+func SetAppTypeId(_ context.Context, value string) error {
+	appTypeIdMutex.Lock()
+	defer appTypeIdMutex.Unlock()
+
+	appTypeId = &value
+	return nil
+}
+
+func Prefix[T Cacheable](value *T) (string, error) {
+	switch any(value).(type) {
+	case *models.Account:
+		return "account", nil
+	case *clients.AccountDetailsAWS:
+		return "account_detail_aws", nil
+	case *clients.AzureTenantId:
+		return "azure_tenant_id", nil
+	default:
+		return "", ErrUnknownType
+	}
+}
+
+func Find[T Cacheable](ctx context.Context, key string, value *T) error {
+	if !redisEnabled {
+		return ErrNotFound
+	}
+
+	prefix, err := Prefix(value)
+	if err != nil {
+		return fmt.Errorf("redis prefix error: %w", err)
+	}
+
+	cmd := client.Get(ctx, prefix+key)
+	if errors.Is(cmd.Err(), redis.Nil) {
+		metrics.IncCacheHit(prefix, "miss")
+		return ErrNotFound
+	} else if cmd.Err() != nil {
+		metrics.IncCacheHit(prefix, "err")
+		return fmt.Errorf("redis get error: %w", cmd.Err())
+	}
+
+	buf, err := cmd.Bytes()
+	if err != nil {
+		metrics.IncCacheHit(prefix, "err")
+		return fmt.Errorf("redis bytes conversion error: %w", err)
+	}
+
+	dec := gob.NewDecoder(bytes.NewReader(buf))
+
+	err = dec.Decode(value)
+	if err != nil {
+		// decode error can be thrown if previous cache entry was JSON-encoded, return not found to overwrite it
+		ctxval.Logger(ctx).Warn().Err(err).Bool("cache", true).Msgf("redis cache decode error: %s", err.Error())
+		metrics.IncCacheHit(prefix, "err")
+		return ErrNotFound
+	}
+
+	metrics.IncCacheHit(prefix, "hit")
+	return nil
+}
+
+func SetExpires[T Cacheable](ctx context.Context, key string, value *T, expiration time.Duration) error {
+	if !redisEnabled {
+		return nil
+	}
+
+	prefix, err := Prefix(value)
+	if err != nil {
+		return fmt.Errorf("redis prefix error: %w", err)
+	}
+
+	if value == nil {
+		metrics.IncCacheHit(prefix, "err")
+		return nil
+	}
+
+	var buf bytes.Buffer
+	enc := gob.NewEncoder(&buf)
+
+	err = enc.Encode(*value)
+	if err != nil {
+		metrics.IncCacheHit(prefix, "err")
+		return fmt.Errorf("unable to encode for Redis cache: %w", err)
+	}
+
+	cmd := client.Set(ctx, prefix+key, buf.String(), expiration)
+	if cmd.Err() != nil {
+		metrics.IncCacheHit(prefix, "err")
+		return fmt.Errorf("redis set error: %w", cmd.Err())
+	}
+
+	return nil
+}
+
+// nolint: wrapcheck
+func SetForever[T Cacheable](ctx context.Context, key string, value *T) error {
+	return SetExpires(ctx, key, value, Forever)
+}
+
+// nolint: wrapcheck
+func Set[T Cacheable](ctx context.Context, key string, value *T) error {
+	return SetExpires(ctx, key, value, config.Application.Cache.Expiration)
+}
```

Comment:
The interface is simple:

* Find
* Set
* SetForever (with a very long expiration)
* SetExpires (with your own expiration duration)

I see I forgot to document these functions, will do during rebase.